trace_replay/trace_replay.cc

```diff
@@ -471,269 +470,167 @@ Status Tracer::WriteTrace(const Trace& trace) {
 
 Status Tracer::Close() { return WriteFooter(); }
 
-Replayer::Replayer(DB* db, const std::vector<ColumnFamilyHandle*>& handles,
-                   std::unique_ptr<TraceReader>&& reader)
-    : trace_reader_(std::move(reader)) {
+ReplayerImpl::ReplayerImpl(DBImpl* db,
+                           const std::vector<ColumnFamilyHandle*>& handles,
+                           std::unique_ptr<TraceReader>&& reader)
+    : Replayer(),
+      trace_reader_(std::move(reader)),
+      prepared_(false),
+      trace_end_(false),
+      header_ts_(0) {
   assert(db != nullptr);
-  db_ = static_cast<DBImpl*>(db->GetRootDB());
+  db_ = db;
   env_ = Env::Default();
   for (ColumnFamilyHandle* cfh : handles) {
     cf_map_[cfh->GetID()] = cfh;
   }
-  fast_forward_ = 1;
 }
 
-Replayer::~Replayer() { trace_reader_.reset(); }
+ReplayerImpl::~ReplayerImpl() { trace_reader_.reset(); }
 
-Status Replayer::SetFastForward(uint32_t fast_forward) {
-  Status s;
-  if (fast_forward < 1) {
-    s = Status::InvalidArgument("Wrong fast forward speed!");
-  } else {
-    fast_forward_ = fast_forward;
-    s = Status::OK();
-  }
-  return s;
-}
-
-Status Replayer::Replay() {
-  Status s;
+Status ReplayerImpl::Prepare() {
   Trace header;
   int db_version;
-  s = ReadHeader(&header);
+  Status s = ReadHeader(&header);
   if (!s.ok()) {
     return s;
   }
   s = TracerHelper::ParseTraceHeader(header, &trace_file_version_, &db_version);
   if (!s.ok()) {
     return s;
   }
+  header_ts_ = header.ts;
+  prepared_ = true;
+  trace_end_ = false;
```

Comment:
Won't these values be set incorrectly if Prepare succeeded the first time and failed the second?